packages/core/streams/streaming-text-response.ts

```diff
@@ -16,29 +16,116 @@ export class StreamingTextResponse extends Response {
+ */
+export function streamToNodeReadable(stream: ReadableStream) {
+  const reader = stream.getReader()
+  let closed = false
+
+  // in order to avoid module check error
+  const Readable = require('stream')
+    .Readable as typeof import('stream').Readable
+
+  const readable = new Readable({
+    read() {
+      reader
+        .read()
+        .then(({ done, value }) => {
+          if (done) {
+            this.push(null)
+          } else {
+            this.push(value)
+          }
+        })
+        .catch(e => {
+          readable.destroy(e)
+        })
+    },
+    destroy(error, callback) {
+      const done = () => {
+        try {
+          callback(error)
+        } catch (err: unknown) {
+          process.nextTick(() => {
+            throw err
+          })
+        }
+      }
+
+      if (!closed) {
+        reader.cancel(error).then(done, done)
+        return
+      }
+      done()
+    }
+  })
+
+  reader.closed.then(
+    () => {
+      closed = true
+    },
+    error => {
+      readable.destroy(error)
+    }
+  )
+
+  return readable
+}
+
+/**
+ * A utility function to get a OutgoingHttpHeaders from a HeadersInit.
+ */
+export function headersInitToOutgoingHeaders(headers: HeadersInit | undefined) {
+  const h = new Headers(headers)
+  const outgoingHeaders: OutgoingHttpHeaders = {}
+  h.forEach((value, key) => {
+    if (
+      key === 'set-cookie' &&
+      'getSetCookie' in h &&
+      typeof h.getSetCookie === 'function'
+    ) {
+      // nodejs 18.14.1 supports it, otherwise developer should polyfill this method for their own
+      outgoingHeaders[key] = h.getSetCookie()
+      return
+    }
+    outgoingHeaders[key] = value
+  })
+  return outgoingHeaders
+}
+
 /**
  * A utility function to stream a ReadableStream to a Node.js response-like object.
  */
 export function streamToResponse(
   res: ReadableStream,
   response: ServerResponse,
-  init?: { headers?: Record<string, string>; status?: number }
+  init?: { headers?: HeadersInit; status?: number }
 ) {
+  if (response.destroyed || response.writableEnded || response.headersSent) {
+    return res.cancel(new Error('Server response is already used'))
+  }
+
+  // start write the response headers
   response.writeHead(init?.status || 200, {
-    'Content-Type': 'text/plain; charset=utf-8',
-    ...init?.headers
+    'content-type': 'text/plain; charset=utf-8',
+    ...headersInitToOutgoingHeaders(init?.headers)
   })
 
-  const reader = res.getReader()
-  function read() {
-    reader.read().then(({ done, value }: { done: boolean; value?: any }) => {
-      if (done) {
-        response.end()
-        return
+  // pipe from node readable to server response
+  // which supports backpresure mechanism
+  const readable = streamToNodeReadable(res)
+
+  const pipeline = require('node:stream')
+    .pipeline as typeof import('node:stream').pipeline
+  return new Promise<void>((resolve, reject) => {
+    pipeline(readable, response, error => {
```

Comment:
We have an [implementation of this in the `edge-runtime` package](https://unpkg.com/edge-runtime@2.4.3/dist/server/body-streams.js) which should be a little faster because it skips the `Readable` indirection.

Reply:
> We have an [implementation of this in the `edge-runtime` package](https://unpkg.com/edge-runtime@2.4.3/dist/server/body-streams.js) which should be a little faster because it skips the `Readable` indirection.

First Pipe in @edge/runtime dose not process possible reader error, which should have destroyed res with that error. Meanwhile it checks open state in every chunk cycle which is not as fast as you think. Third it is another package, and do we really need to import another or just leave this method alone?

For me, I think we should provide a nodejs stream conversion here for the ai sdk implements so many streams which is helpful and users can easily deploy their apps with only this package in nodejs.

Reply:
> First Pipe in @edge/runtime dose not process possible reader error, which should have destroyed res with that error

The `pipeBodyStreamToResponse` is designed so that the caller handles closing the response (in case it wants to send more data). We just need a `.then(destroy, destroy)` to handle.

> Meanwhile it checks open state in every chunk cycle which is not as fast as you think.

`open` is a local variable, so I'm not sure what you mean.

> Third it is another package, and do we really need to import another or just leave this method alone?

This is a Vercel library consuming another Vercel library, and I think commonly used inside Next.js (which will have already loaded the `edge-runtime` library). In that case, changing will actually reduce code at runtime. And if this isn't being used by a Next.js server, a bundler will be able to strip out everything not used by us.